Derive does not work with const generics

```
error[E0747]: constant provided when a type was expected
   --> tests/deriving.rs:197:35
    |
197 | enum EnumWithGeneric<'a, T, const N: usize> {
    |                                   ^

error[E0392]: type parameter `N` is never used
   --> tests/deriving.rs:197:35
    |
197 | enum EnumWithGeneric<'a, T, const N: usize> {
    |                                   ^
    |                                   |
    |                                   unused type parameter
    |                                   `N` is named here, but is likely unused in the containing type
    |
    = help: consider removing `N`, referring to it in a field, or using a marker such as `PhantomData`
    = help: if you intended `N` to be a const parameter, use `const N: /* Type */` instead
```

In the generated helper types, we're failing to declare the const generic parameter as a const generic rather than a type generic.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+### Added
+
+* `derive(Exhaust)` now supports types with const generic parameters.
+
 ### Changed
 
 * `derive(Exhaust)` generates simpler code to improve build performance (and readability).

--- a/exhaust-macros/src/lib.rs
+++ b/exhaust-macros/src/lib.rs
@@ -237,7 +237,7 @@ fn exhaust_iter_struct(
 ) -> Result<(TokenStream2, TokenStream2), syn::Error> {
     let vis = &ctx.vis;
     let exhaust_crate_path = &ctx.exhaust_crate_path;
-    let (impl_generics, ty_generics, augmented_where_predicates) =
+    let (impl_or_decl_generics, ty_generics, augmented_where_predicates) =
         ctx.generics_with_bounds(syn::parse_quote! {});
     let iterator_type_name = &ctx.iterator_type_name;
     let factory_type_name = &ctx.factory_type.path()?;
@@ -303,13 +303,13 @@ fn exhaust_iter_struct(
         // This struct is always wrapped in a newtype struct to hide implementation details reliably.
         let factory_state_struct_decl = match &factory_field_decls {
             syn::Fields::Unit | syn::Fields::Unnamed(_) => quote! {
-                #vis struct #factory_state_struct_type #ty_generics #factory_field_decls
+                #vis struct #factory_state_struct_type #impl_or_decl_generics #factory_field_decls
                 where #augmented_where_predicates;
 
             },
 
             syn::Fields::Named(_) => quote! {
-                #vis struct #factory_state_struct_type #ty_generics
+                #vis struct #factory_state_struct_type #impl_or_decl_generics
                 where #augmented_where_predicates
                 #factory_field_decls
             },
@@ -363,14 +363,16 @@ fn exhaust_iter_struct(
             //
             // Note: The iterator struct must have trait bounds because its fields, being of type
             // `<SomeOtherTy as Exhaust>::Iter`, require that `SomeOtherTy: Exhaust`.
-            #vis struct #iterator_type_name #ty_generics
+            #vis struct #iterator_type_name #impl_or_decl_generics
             where #augmented_where_predicates {
                 #state_field_decls
             }
 
             // Struct that is exposed as the `<Self as Exhaust>::Factory` type,
             // wrapping the private factory_state_struct_type.
-            #vis struct #factory_type_name #ty_generics (#factory_state_struct_type #ty_generics)
+            #vis struct #factory_type_name #impl_or_decl_generics (
+                #factory_state_struct_type #ty_generics
+            )
             where #augmented_where_predicates;
 
             #impls
@@ -381,7 +383,7 @@ fn exhaust_iter_struct(
             #factory_state_struct_decl
 
             // A manual impl of Clone is required to *not* have a `Clone` bound on the generics.
-            impl #impl_generics ::core::clone::Clone for #factory_type
+            impl #impl_or_decl_generics ::core::clone::Clone for #factory_type
             where #augmented_where_predicates {
                 fn clone(&self) -> Self {
                     Self(#factory_state_struct_clone_expr)

--- a/tests/deriving.rs
+++ b/tests/deriving.rs
@@ -100,13 +100,12 @@ fn struct_simple() {
     )
 }
 
+/// A struct with type, lifetime, and const parameters, and a trait bound on the type parameter.
 #[derive(Debug, exhaust::Exhaust, PartialEq)]
-struct GenericStruct<'a, T> {
+struct GenericStruct<'a, T: std::marker::Copy, const N: usize> {
     a: T,
     b: T,
-    p: std::marker::PhantomData<&'a ()>,
-    // TODO: Also validate that trait bounds on the struct are handled
-    // TODO: Test with const generics.
+    p: std::marker::PhantomData<&'a [(); N]>,
 }
 
 #[test]
@@ -114,7 +113,7 @@ fn struct_generic() {
     let p = std::marker::PhantomData;
     #[cfg_attr(any(), rustfmt::skip)]
     std::assert_eq!(
-        c::<GenericStruct<bool>>(),
+        c::<GenericStruct<bool, 3>>(),
         std::vec![
             GenericStruct { a: false, b: false, p },
             GenericStruct { a: false, b: true, p },
@@ -131,7 +130,10 @@ struct UninhabitedStruct {
 
 #[test]
 fn struct_uninhabited_generic() {
-    std::assert_eq!(c::<GenericStruct<std::convert::Infallible>>(), std::vec![])
+    std::assert_eq!(
+        c::<GenericStruct<std::convert::Infallible, 100>>(),
+        std::vec![]
+    )
 }
 
 #[test]

